vpr/src/place/initial_placement.cpp

```diff
@@ -0,0 +1,426 @@
+
+            // Try to place the macro first, if can be placed - place them, otherwise try again
+            for (itry = 0; itry < macros_max_num_tries && macro_placed == false; itry++) {
+                // Choose a random position for the head
+                ipos = vtr::irand(free_locations[itype] - 1);
+
+                // Try to place the macro
+                macro_placed = try_place_macro(itype, ipos, pl_macro);
+
+            } // Finished all tries
+
+            if (macro_placed == false) {
+                // if a macro still could not be placed after macros_max_num_tries times,
+                // go through the chip exhaustively to find a legal placement for the macro
+                // place the macro on the first location that is legal
+                // then set macro_placed = true;
+                // if there are no legal positions, error out
+
+                // Exhaustive placement of carry macros
+                for (ipos = 0; ipos < free_locations[itype] && macro_placed == false; ipos++) {
+                    // Try to place the macro
+                    macro_placed = try_place_macro(itype, ipos, pl_macro);
+
+                } // Exhausted all the legal placement position for this macro
+
+                // If macro could not be placed after exhaustive placement, error out
+                if (macro_placed == false) {
+                    // Error out
+                    VPR_FATAL_ERROR(VPR_ERROR_PLACE,
+                                    "Initial placement failed.\n"
+                                    "Could not place macro length %zu with head block %s (#%zu); not enough free locations of type %s (#%d).\n"
+                                    "Please manually size the FPGA because VPR can't do this yet.\n",
+                                    pl_macro.members.size(), cluster_ctx.clb_nlist.block_name(blk_id).c_str(), size_t(blk_id), device_ctx.physical_tile_types[itype].name, itype);
+                }
+
+            } else {
+                // This macro has been placed successfully, proceed to place the next macro
+                continue;
+            }
+        }
+    } // Finish placing all the pl_macros successfully
+}
+
+/* Place blocks that are NOT a part of any macro.
+ * We'll randomly place each block in the clustered netlist, one by one. */
+static void initial_placement_blocks(int* free_locations, enum e_pad_loc_type pad_loc_type) {
+    int itype, ipos;
+    auto& cluster_ctx = g_vpr_ctx.clustering();
+    auto& place_ctx = g_vpr_ctx.mutable_placement();
+
+    // The map serves to place first the most constrained block ids
+    std::map<int, std::vector<ClusterBlockId>> sorted_block_map;
+
+    for (auto blk_id : cluster_ctx.clb_nlist.blocks()) {
+        auto logical_block = cluster_ctx.clb_nlist.block_type(blk_id);
+
+        size_t num_equivalent_tiles = logical_block->equivalent_tiles.size();
+        sorted_block_map[num_equivalent_tiles].push_back(blk_id);
+    }
```

Comment:
It would likely be clearer and more straight forward to extend in the future if we just had an plain vector of ClusterBlockId, which we then sort according to some criteria:

```cpp
auto blocks = cluster_ctx.clb_nlist.blocks();
std::vector<ClusterBlockId> sorted_blocks(blocks.begin(), blocks.end());

auto criteria = [](const ClusterBlockId lhs, ClusterBlockId rhs) {
  //Sort by most possible instances in placement grid
}

std::sort(sorted_blocks.begin(), sorted_blocks.end(), criteria);
```
This makes it straight forward to change the criteria in the future (rather than requiring that it be sorted by number of tiles as currently implemented).

It also makes it possible to extend this initial placement algorithm with the 'push to front heuristic' for difficult to place blocks as previously discussed (simply move problematic blocks to the front of the list and re-do initial placement).

Reply:
I have added the sorting step instead of the `map` solution. The only problem with this, as far as I have locally experienced, is that the regression tests will complete, but some of the QoR results will be different. I have checked the placer outputs and they are actually different and the run-time has slightly increased (for what regards strong tests).

Would it be ok, in your opinion, to add a check to see whether the architecture does have equivalent sites or not. In the case it has some tiles with multiple sites, the sorting is triggered, otherwise not.

Reply:
Since this would change the initial placement it means the placer would take a different walk through the space of possible placements, so I would expect the final placements (and hence QoR) to change somewhat.

So long as the new QoR is reasonable (e.g. on the VTR and Titan benchmarks) I'm OK if it changes.

Reply:
Ok, once I have performed all the changes I will go through another Titan benchmarks run and see the outcome.

There is one failure though of a strong regression test in CI which cannot route within 50 iterations: https://travis-ci.org/verilog-to-routing/vtr-verilog-to-routing/jobs/614569674?utm_medium=notification&utm_source=github_status. This seems odd, but I need to investigate what is happening there